File: apps/mobile/test/claims_test.dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

/// Claim discipline for the mobile app, as an executable check rather than
/// prose — the marketing site's claims.test.ts pattern, extended to the field
/// app (P0-PRODUCT-VISIBILITY-002).
///
/// The capability-visibility audit found ZERO overclaims in lib/ — the app's
/// honesty is structural (mock hardware is compiled out of release builds and
/// refused by the platform in production; "QR scanning arrives with device
/// integration" and "no PDF engine yet" label the future as future). This test
/// makes that honesty hold by construction: a string that asserts an
/// unavailable capability fails the suite before it ships.
///
/// Unlike the portal, the mobile app has NO labelled roadmap surface, so there
/// is no allowed home for this vocabulary at all — every pattern below is
/// banned outright. Deliberately NOT banned: "mock scale"/"mock analyzer"
/// (honest dev-only labels), the rendered supplier QR (real — only SCANNING is
/// future), "QR scanning arrives with device integration" (labelled future),
/// and comments naming the web portal (a real product).
class _Claim {
  _Claim(this.pattern, this.why);
  final RegExp pattern;
  final String why;
}

void main() {
  final claims = <_Claim>[
    // -- AI: mobile never names AI at all; the deviation flag lives (and is
    //    described) on the platform, never on this UI or the parchi.
    _Claim(RegExp(r'\bAI\b'), 'no AI exists in the product; mobile never names it'),
    _Claim(RegExp(r'machine[- ]learning', caseSensitive: false), 'no ML is deployed'),
    _Claim(RegExp(r'artificial intelligence', caseSensitive: false), 'no ML is deployed'),
    _Claim(RegExp(r'predict', caseSensitive: false), 'nothing predicts; forecasting is a roadmap item'),
    _Claim(RegExp(r'\bforecast', caseSensitive: false), 'forecasting is a V2 roadmap item, not built'),
    // -- Location: no GPS/location code or dependency exists.
    _Claim(RegExp(r'\bGPS\b'), 'no GPS exists and it is never a pilot dependency'),
    _Claim(RegExp(r'location track|geofenc', caseSensitive: false), 'no location tracking exists'),
    // -- Enterprise: reserved and not built; the field app must never imply it.
    _Claim(RegExp(r'\bSAP\b'), 'SAP/ERP integration is ENTERPRISE roadmap - no vendor, no protocol'),
    _Claim(RegExp(r'\bSSO\b|single sign-?on', caseSensitive: false), 'enterprise SSO is not built'),
    _Claim(RegExp(r'federat(ion|ed)', caseSensitive: false), 'federation/org-to-org is not built'),
    _Claim(RegExp(r'global identity', caseSensitive: false), 'global identity is not built'),
    // -- Apps that do not exist. ("customer app" is real - a household screen
    //    ships in this binary - so only the farmer app and the future WEB
    //    outlet portal are banned.)
    _Claim(RegExp(r'farmer app', caseSensitive: false), 'there is no farmer app; farmers are records who receive a parchi'),
    _Claim(RegExp(r'outlet portal', caseSensitive: false), 'the web outlet portal is a future option, not built'),
    // -- Messaging: no BSP/DLT provider is contracted; nothing sends.
    _Claim(RegExp(r'whatsapp', caseSensitive: false), 'no messaging provider is contracted; the parchi is shared as plain text'),
    _Claim(RegExp(r'(sent|delivered) (via|over|by) SMS', caseSensitive: false), 'no SMS provider is contracted'),
    // -- Hardware: capture is manual-first; mocks are dev-only and refused in
    //    production. Naming the mock honestly stays allowed; claiming a live
    //    device does not.
    _Claim(RegExp(r'automatically (reads?|captures?|weighs?|measures?)', caseSensitive: false), 'capture is manual; automated read-assist is discovery-gated roadmap'),
    _Claim(RegExp(r'(scale|analyzer) (is )?(connected|integrated|online)', caseSensitive: false), 'no device integration is shipped'),
    _Claim(RegExp(r'reads? the (scale|analyzer)', caseSensitive: false), 'no device integration is shipped'),
    _Claim(RegExp(r'\bIoT\b', caseSensitive: false), 'no IoT capability exists'),
    // -- QR: rendering the supplier QR is real; scanning one is future.
    _Claim(RegExp(r'scan (a|the|your) (QR|code|barcode)|tap to scan', caseSensitive: false), 'QR scanning is not built; only rendering a QR is real'),
    // -- Government / compliance: documents, not integrations; no attestations.
    _Claim(RegExp(r'government (integration|portal|filing|approved)', caseSensitive: false), 'no government integration exists'),
    _Claim(RegExp(r'SOC ?2|ISO[- ]?(?!8601|4217)\d{4,5}|GDPR|HIPAA|PCI[- ]DSS', caseSensitive: false), 'no certifications or compliance attestations exist to claim (ISO 8601 dates and ISO 4217 currency codes are factual standards, not attestations)'),
    _Claim(RegExp(r'(legally|fully) compliant|100% secure', caseSensitive: false), 'no compliance or absolute-security claims'),
  ];

  final files = Directory('lib')
      .listSync(recursive: true)
      .whereType<File>()
      .where((f) => f.path.endsWith('.dart'))
      .toList();

  test('finds the source tree', () {
    expect(files.length, greaterThan(10));
  });

  for (final claim in claims) {
    test('never says ${claim.pattern.pattern} (${claim.why})', () {
      final offenders = <String>[
        for (final file in files)
          if (claim.pattern.hasMatch(file.readAsStringSync())) file.path,
      ];
      expect(offenders, isEmpty,
          reason: 'overclaim "${claim.pattern.pattern}" found in: '
              '${offenders.join(", ")} - ${claim.why}');
    });
  }
}

File: apps/admin-portal/src/app/claims.test.ts
import { readdirSync, readFileSync, statSync } from "node:fs";
import { join, relative } from "node:path";
import { describe, expect, it } from "vitest";

/**
 * Claim discipline for the portal, as an executable check rather than prose —
 * the marketing site's claims.test.ts pattern, extended to product UI
 * (P0-PRODUCT-VISIBILITY-002).
 *
 * The capability-visibility audit (LACTEVA-PRODUCT-CAPABILITY-VISIBILITY-AUDIT)
 * found ZERO overclaims in this tree — honesty held by inspection. This test
 * makes it hold by construction: copy that asserts an unavailable capability
 * (AI beyond the statistical deviation flag, GPS, SAP, SSO, automated
 * scale/analyzer capture, QR scanning, WhatsApp/SMS sending, forecasting,
 * federation, farmer app, web outlet portal, compliance attestations) fails
 * the suite before it ships.
 *
 * Two tiers, because honesty has two shapes here:
 *  - FORBIDDEN: claim shapes that cannot be honest anywhere today.
 *  - ROADMAP_ONLY: vocabulary that IS legitimate — but only on the roadmap
 *    page, where every item is explicitly labelled Coming soon / Enterprise /
 *    Future and rendered inert (see roadmap-page.test.tsx). Anywhere else in
 *    the portal, naming these capabilities reads as having them.
 *
 * Deliberately NOT banned: the honest anti-claims this tree already carries —
 * "nothing here pretends a device supplied a value", "not that WhatsApp will
 * reach it", "does not try to predict that" — and factual features that share
 * a noun with a future one (the supplier QR is really RENDERED today; only
 * scanning is future).
 */

const ROADMAP_DIR = join("app", "roadmap");

const FORBIDDEN: Array<{ pattern: RegExp; why: string }> = [
  // -- AI: the only intelligence shipped is a statistical deviation flag.
  { pattern: /(?<!before )AI-powered/i, why: "no ML is deployed; the deviation flag is statistics" },
  { pattern: /powered by (AI|artificial intelligence)/i, why: "no ML is deployed" },
  { pattern: /machine[- ]learning (predicts|detects|powers|drives)/i, why: "no ML is deployed" },
  { pattern: /predicti(ve|on)/i, why: "nothing predicts; anomaly/forecasting are roadmap items" },
  // -- Location: no GPS/location code exists anywhere in the product.
  { pattern: /\bGPS\b/, why: "no GPS exists and it is never a pilot dependency" },
  { pattern: /location track|geofenc/i, why: "no location tracking exists" },
  // -- Hardware: capture is manual-first; mock readings are refused in prod.
  { pattern: /automatically (reads?|captures?|weighs?|measures?)/i, why: "capture is manual; automated read-assist is discovery-gated roadmap" },
  { pattern: /(scale|analyzer) (is )?(connected|integrated|online)/i, why: "no device integration is shipped" },
  { pattern: /reads? the (scale|analyzer)/i, why: "no device integration is shipped" },
  { pattern: /\bIoT\b/i, why: "no IoT capability exists" },
  // -- QR: the supplier QR is rendered (real); scanning it is roadmap.
  { pattern: /scan (a|the|your) (QR|code|barcode)|tap to scan/i, why: "QR scanning is not built; only rendering a QR is real" },
  // -- Messaging: templates and registry are real; no BSP/DLT provider is
  //    contracted, so nothing can promise a send.
  { pattern: /WhatsApp (is )?(connected|enabled|live)\b/i, why: "no messaging provider is contracted" },
  { pattern: /(sent|delivered) (via|over|by) (WhatsApp|SMS)/i, why: "no messaging provider is contracted; nothing is sent today" },
  // -- Government / compliance: documents, not integrations; no attestations.
  { pattern: /government (integration|portal|filing|approved)/i, why: "no government integration exists" },
  { pattern: /SOC ?2|ISO[- ]?(?!8601|4217)\d{4,5}|GDPR|HIPAA|PCI[- ]DSS/i, why: "no certifications or compliance attestations exist to claim (ISO 8601 dates and ISO 4217 currency codes are factual standards, not attestations)" },
  { pattern: /(legally|fully) compliant|100% secure|absolutely secure/i, why: "no compliance or absolute-security claims" },
];

const ROADMAP_ONLY: Array<{ pattern: RegExp; why: string }> = [
  { pattern: /\bAI\b/, why: "AI may be named only where it is labelled not-built (the roadmap page)" },
  { pattern: /machine[- ]learning/i, why: "ML exists only as a labelled roadmap item" },
  { pattern: /artificial intelligence/i, why: "AI exists only as a labelled roadmap item" },
  { pattern: /\bSAP\b/, why: "SAP/ERP integration is ENTERPRISE roadmap — no vendor, no protocol" },
  { pattern: /\bSSO\b|single sign-?on/i, why: "enterprise SSO is ENTERPRISE roadmap, not built" },
  { pattern: /federat(ion|ed)/i, why: "federation/org-to-org is ENTERPRISE roadmap, not built" },
  { pattern: /global identity/i, why: "global identity is ENTERPRISE roadmap, not built" },
  { pattern: /\bforecast/i, why: "forecasting is a V2 roadmap item, not built" },
  { pattern: /farmer app/i, why: "there is no farmer app; farmers are records who receive a parchi" },
  { pattern: /outlet portal/i, why: "the web outlet portal is a future option, not built" },
];

function collectSourceFiles(dir: string): string[] {
  const files: string[] = [];
  for (const entry of readdirSync(dir)) {
    const path = join(dir, entry);
    if (statSync(path).isDirectory()) {
      files.push(...collectSourceFiles(path));
    } else if (/\.(ts|tsx)$/.test(entry) && !entry.includes(".test.")) {
      files.push(path);
    }
  }
  return files;
}

describe("claim discipline", () => {
  const root = join(__dirname, "..");
  const files = collectSourceFiles(root);

  it("finds the source tree", () => {
    expect(files.length).toBeGreaterThan(10);
  });

  for (const { pattern, why } of FORBIDDEN) {
    it(`never says ${pattern} (${why})`, () => {
      const offenders = files.filter((file) =>
        pattern.test(readFileSync(file, "utf8")),
      );
      expect(offenders).toEqual([]);
    });
  }

  for (const { pattern, why } of ROADMAP_ONLY) {
    it(`says ${pattern} only on the labelled roadmap page (${why})`, () => {
      const offenders = files.filter(
        (file) =>
          !relative(root, file).startsWith(ROADMAP_DIR) &&
          pattern.test(readFileSync(file, "utf8")),
      );
      expect(offenders).toEqual([]);
    });
  }
});
